Add booking_rules.json (#11)
GOFS-lite/GOFS-lite: reference.md

`operating_rules.json` | REQUIRED | Defines rules for intra-zone and inter-zone trips as well as operating times.
`calendars.json` | REQUIRED | Defines dates and days when on-demand services are available to the riders.
`fares.json` | OPTIONAL | Defines static fare rules for a system. 
`wait_times.json` | Optionally REQUIRED | Defines global wait time for defined areas of service. Either `wait_times.json` or `wait_time` or `booking_rules.json` MUST be provided.
`wait_time` | Optionally REQUIRED | Returns a wait time for queried areas. Either `wait_times.json` or `wait_time` or `booking_rules.json` MUST be provided.

> `booking_rules.json` | Optionally REQUIRED | Returns a wait time for queried areas. Either `wait_times.json` or `wait_time` or `booking_rules.json` MUST be provided.

Reviewer: ```suggestion
`booking_rules.json` | Optionally REQUIRED | Returns rules for booking in queried areas when live ETA or real time booking is not available. Either `wait_times.json` or `wait_time` or `booking_rules.json` MUST be provided.
```

  }
}
```

### booking_rules.json

> This file defines rules about how to book a ride in advance and is to be used instead `wait_times` if users can not order a ride in real-time. With ride bookings, no wait times are applied. If available, users can either book a ride for the same day with an advance notice or for a futur day. 

Author: Agreed for the first point, those fields are helpful. 

Ambivalent on the second point. 
* At the minimum, `wait_times.geojson` provides static wait time. It'd be a weird situation to be able to book for a specific time, but assuming at the same time that the ride will always have 15 minutes delay. 
* For `wait_times`, that provides real-time wait times, shouldn't your wait time for a pre-book ride be specific to your booking, and not a "generic" wait times? There's a difference between a wait time between two locations, and a wait time between two locations for which you have a reservation already. I am not sure how to express this however. 

> but at the same time may face delays to pick up the riders on time (eg: facing a 2 hours' delay, because of occasional high ridership demand).
I also assume in general that agencies pre-booking feature will try to first confirm that they don't overbook their vehicles either. So a surge in demand would simply results in unsuccessful pre-booking for the user.

> This file defines rules about how to book a ride in advance and is to be used instead `wait_times` if users can not order a ride in real-time. With ride bookings, no wait times are applied. If available, users can either book a ride for the same day with an advance notice or for a futur day. 

Reviewer: What you say is true in theory @timMillet but I'm not sure that's the case in practice. Either you have a wait time with real time booking or you book for later. I've never seen both concept mixed.

> This file defines rules about how to book a ride in advance and is to be used instead `wait_times` if users can not order a ride in real-time. With ride bookings, no wait times are applied. If available, users can either book a ride for the same day with an advance notice or for a futur day. 

Reviewer 2: The PR as it makes sense if GOFS-lite is a stand-alone; in a real-time feed, wait time and booking time can overlap. The issue that I see is mostly related to compatibility with GTFS-Flex and GTFS-OnDemand, more specifically booking_rules.txt and wait_rules.txt.
In these static files, there is nothing that forbid an on-demand service operator to provide a general booking rule valid at all times, and wait times that differ depending on dates, days, time, and locations. How would we translate that using the spec from this PR?

| Field Name | Presence | Type |Description |
 ---------- | ---- | -------- | ----------- |
 `booking_rules` |  REQUIRED  | Array | Array that contains one object per booking rules as defined below. |
\-&nbsp;`from_s2_cells` | Conditionally REQUIRED | Array | The reference to one or many S2CellID that cover the area of the wait time update. Information on S2 cells can be found here https://s2geometry.io/. Required if `from_zone_ids` field is not populated. Forbidden otherwise.
\-&nbsp;`to_s2_cells` | OPTIONAL | Array | The reference to one or many S2CellID that cover the area of the destination. Information on S2 cells can be found here https://s2geometry.io/. Optional if `from_s2_cells` field is populated. Forbidden otherwise.
\-&nbsp;`from_zone_ids` | Conditionally REQUIRED | Array | One or many ID from a zone defined in `zones.json`  that cover the area of the wait time update. Required if `from_s2_cells` field is not populated. Forbidden otherwise.
\-&nbsp;`to_zone_ids` | OPTIONAL | Array | One or many ID from a zone defined in `zones.json`  that cover the area of the destination. Optional if `from_zone_ids` field is populated. Forbidden otherwise.

>  \-&nbsp; `booking_type` | REQUIRED | Enum | Indicates how far in advance booking can be made. <br><br>Valid options are:<br>`0` - Up to same-day booking with advance notice.<br>`1` - Up to prior day(s) booking. |

Reviewer: ```suggestion
 \-&nbsp; `booking_type` | REQUIRED | Enum | Indicates how far in advance booking can be made. <br><br>Valid options are:<br>`0` - Up to same-day booking with advance notice.<br>`1` - Up to prior day(s) booking. <br><br> If real time booking is available, please use either `wait_times.json` or `wait_time` instead. |
```

>  \-&nbsp; `booking_type` | REQUIRED | Enum | Indicates how far in advance booking can be made. <br><br>Valid options are:<br>`0` - Up to same-day booking with advance notice.<br>`1` - Up to prior day(s) booking.<br><br>If real time booking is available, please use either `wait_times.json` or `wait_time` instead. ||

Reviewer: ```suggestion
 \-&nbsp; `booking_type` | REQUIRED | Enum | Indicates how far in advance booking can be made. <br><br>Valid options are:<br>`same-day` - Up to same-day booking with advance notice.<br>`prior` - Up to prior day(s) booking.<br><br>If real time booking is available, please use either `wait_times.json` or `wait_time` instead. ||
```

Not sure if those are the right values? But at least it should not be 0/1. 0/1 in GTFS is because CSVs

Reviewer 2: I don't think what is the booking_type=0 (ie realtime booking) in GTFS should be removed; see comment above.
If an enum is used, please use the same values as in GTFS (to prevent any source of confusion) even though one of them is removed.

>  \-&nbsp; `prior_notice_duration_min` | Conditionally REQUIRED | Integer | Minimum number of minutes before travel to make the request. REQUIRED for `booking_type=0`. Forbidden otherwise. |

Reviewer: ```suggestion
 \-&nbsp; `prior_notice_duration_min` | Conditionally REQUIRED | Integer | Minimum number of minutes before travel to make the request. REQUIRED for `booking_type=0`. FORBIDDEN otherwise. |
```

Author: I've spotted a few other lower case `Forbidden`, I'll replace them too.

 \-&nbsp; `prior_notice_last_time` | Conditionally REQUIRED | Time | Last time on the last day before travel to make the booking request (e.g. “Ride must be booked 1 day in advance before 5PM” will be encoded as `prior_notice_last_time=17:00:00`). REQUIRED if `prior_notice_last_day` is defined. FORBIDDEN otherwise. |
 \-&nbsp; `prior_notice_start_day` | OPTIONAL | Integer | Earliest day before travel to make the booking request (e.g.: “Ride can be booked at the earliest one week in advance at midnight” will be encoded as `prior_notice_start_day=7`). FORBIDDEN for `booking_type=0` if `prior_notice_duration_max` is defined. OPTIONAL otherwise. |
 \-&nbsp; `prior_notice_start_time` | Conditionally REQUIRED | Time | Earliest time on the earliest day before travel to make the booking request (e.g. : “Ride can be booked at the earliest one week in advance at midnight” will be encoded as `prior_notice_start_time=00:00:00`). REQUIRED if `prior_notice_start_day` is defined. FORBIDDEN otherwise. |

>  \-&nbsp; `prior_notice_service_id` | OPTIONAL | ID referencing `calendar.service_id` | Indicates the service days on which `prior_notice_last_day` or `prior_notice_start_day` are counted (e.g. : If empty, `prior_notice_start_day=2` will be two calendar days in advance. If defined as a `service_id` containing only business days (weekdays without holidays), `prior_notice_start_day=2` will be two business days in advance). OPTIONAL if `booking_type=1`. FORBIDDEN otherwise. |

Reviewer 2: "ID referencing `calendar.service_id`" is a GTFS way of referencing. See GOFS field `calendars` in `operating_rules.json`.

>  \-&nbsp; `message` | OPTIONAL | String | Message to riders utilizing service at a `stop_time` when booking on-demand pickup and drop off. Meant to provide minimal information to be transmitted within a user interface about the action a rider must take in order to utilize the service. |

Reviewer 2: `stop_time` is a GTFS concept. It could be removed and not replaced.
